### PR TITLE
cql3: Enforce Key-length limits during SELECT

### LINKED ISSUE #10366: Enforce Key-length limits during SELECT
Partition keys length are limited to 64KB, and and clustering keys are similarly limited (the CQL column name needs to fit in 64KB - see issue #3017)

If the user attempts a SELECT based on the **partition key**, i.e., `WHERE p = ?`, if the string is too long Cassandra returns a clean InvalidRequest error with the message "Key length of 66560 is longer than maximum of 65535". In Scylla, we get an unclean internal server error (which looks like a failure to achieve quorum), with the message "Key size too large: 66562 > 65535" (note, by the way, two bytes extra in the message). It should be fixed to give a clean InvalidRequest error.

This behavior is tested in the translated Cassandra unit test `cassandra_tests/validation/operations/select_test.py`::`cassandra_tests/validation/operations/select_test.py`.

Additionally, if selecting based on the **clustering key**, i.e., `WHERE p = 'foo' AND c = ?`, Cassandra is silent (it doesn't find anything, but doesn't return an error either), while Scylla returns the same ugly internal server error described above. I'm not sure why Cassandra chose not to make this into an error, but this behavior is enshrined in their unit test, `cassandra_tests/validation/operations/select_test.py`::`testCKQueryWithValueOver64K`.

### LINKED ISSUE #12247: Better error reporting for oversized keys during INSERT
This issue is very similar to #10366, except that issue is about SELECT, this one is about INSERT.

Partition keys length are limited to 64KB, and and clustering keys are similarly limited (the CQL column name needs to fit in 64KB - see issue https://github.com/scylladb/scylladb/issues/3017)

If the user attempts a INSERT with a partition key or clustering key larger than that limit, Cassandra returns a clean InvalidRequest error with the message "Key length of 66560 is longer than maximum of 65535". In Scylla, we get an unclean internal server error (which looks like a failure to achieve quorum), with the message "Key size too large: 66562 > 65535" (note, by the way, two bytes extra in the reported size, I don't know why). It should be fixed to give a clean `InvalidRequest` error.

This behavior is tested in the translated Cassandra unit test `cassandra_tests/validation/operations/insert_test.py`::`testPKInsertWithValueOver64K`.

### PR DESCRIPTION
**Pull request by ptrsmrn**
Validate if Partition keys length is less then 64k Return InvalidRequest error according to cassandra style in case of the key is too large.

fixes #10366
fixes #12247

Minor, no need to backport.

### CODE REVIEW

**Comment by ptrsmrn** [comment]
This is a continuation of https://github.com/scylladb/scylladb/pull/16012. Per https://github.com/scylladb/scylladb/pull/18407#issuecomment-2077536085, the contributor has already signed the CLA.

File: cql3/restrictions/statement_restrictions.cc [hunk 1]
```diff
@@ -1870,8 +1870,18 @@ dht::partition_range_vector partition_ranges_from_token(const expr::expression&
     return {{std::move(start), std::move(end)}};
 }
 
+void validate_key_length(const std::vector<managed_bytes>& key) {
+    auto max_length = std::numeric_limits<compound_type<allow_prefixes::yes>::size_type>::max();
```
Comment (nuivall) [inline cql3/restrictions/statement_restrictions.cc:1874]: constexpr?

File: test/cqlpy/cassandra_tests/validation/operations/select_test.py [hunk 2]
```diff
@@ -1305,7 +1304,7 @@ def testPKQueryWithValueOver64K(cql, test_keyspace):
                            "SELECT * FROM %s WHERE a = ?", too_big)
 
 # Reproduces #10366
-@pytest.mark.xfail(reason="#10366 - server error instead of silent return")
+@pytest.mark.xfail(reason="Cassandra silently ignores oversized clustering key, Scylla chose to generate an error as in the case of an oversized partition key - see discussion in #10366")
```
Comment (nuivall) [inline test/cqlpy/cassandra_tests/validation/operations/select_test.py:1307]: `Scylla chose to generate an error` - but this test doesn't seem to check that

File: test/cqlpy/cassandra_tests/validation/operations/select_test.py [hunk 3]
```diff
@@ -1296,7 +1296,6 @@ def testIndexQueryWithValueOver64K(cql, test_keyspace):
         assert_empty(execute(cql, table, "SELECT * FROM %s WHERE c = ?  ALLOW FILTERING", too_big))
 
 # Reproduces #10366
-@pytest.mark.xfail(reason="#10366 - server error instead of InvalidRequest")
 def testPKQueryWithValueOver64K(cql, test_keyspace):
     with create_table(cql, test_keyspace, "(a text, b text, PRIMARY KEY (a, b))") as table:
         TOO_BIG = 1024 * 65
```
Comment (nuivall) [inline test/cqlpy/cassandra_tests/validation/operations/select_test.py:1301]: I think we need a better test now, checking both positive and negative scenario (or 2 tests)
- MAX allowed
- MAX allowed + 1

File: cql3/restrictions/statement_restrictions.cc [hunk 4]
```diff
@@ -1870,8 +1870,18 @@ dht::partition_range_vector partition_ranges_from_token(const expr::expression&
...
+    for (auto k : key) {
+        if (k.size() > max_length) {
+            throw exceptions::invalid_request_exception(format("Key length of {} is longer than maximum of {}", k.size(), max_length));
+        }
+    }
+}
+
 /// Turns a partition-key value into a partition_range. \p pk must have elements for all partition columns.
 dht::partition_range range_from_bytes(const schema& schema, const std::vector<managed_bytes>& pk) {
+    validate_key_length(pk);
```
Comment (nuivall) [inline cql3/restrictions/statement_restrictions.cc:1884]: why check it here? it looks like a bit random place

**Comment by ptrsmrn** [comment]
@nuivall thanks for the comments. I have not changed the original commit, just rebased it on top of the latest `master` and verified test.py is passing (I see 1 dtest remains to be fixed). BTW. Nadav approved the original PR with the commit I cherry-picked here. Not sure how to proceed, if I edit the commit I guess the original author won't be mentioned anymore.

File: test/cqlpy/cassandra_tests/validation/operations/select_test.py (hunk 2 above)
Comment (ptrsmrn) [reply test/cqlpy/cassandra_tests/validation/operations/select_test.py:1307]: I think these are tests @nyh rewrote 1:1 from Cassandra's repository, not sure we should change them.

File: cql3/restrictions/statement_restrictions.cc (hunk 4 above)
Comment (ptrsmrn) [reply cql3/restrictions/statement_restrictions.cc:1884]: I believe it's some sort of a common place, but if you have better ideas, I'm all ears.

File: test/cqlpy/cassandra_tests/validation/operations/select_test.py (hunk 2 above)
Comment (nuivall) [reply test/cqlpy/cassandra_tests/validation/operations/select_test.py:1307]: yes I think you're right

File: test/cqlpy/cassandra_tests/validation/operations/select_test.py (hunk 3 above)
Comment (nuivall) [reply test/cqlpy/cassandra_tests/validation/operations/select_test.py:1301]: it tests
LONG_TEXT_LENGTH = 8193
OVERSIZE_LENGTH = 66536

but not OVERSIZE_LENGTH-1

Comment (ptrsmrn) [reply test/cqlpy/cassandra_tests/validation/operations/select_test.py:1301]: If I extended `test_local_indexed_values` dtest with `OVERSIZE_LENGTH-1` case, would that be sufficient?

Comment (ptrsmrn) [reply test/cqlpy/cassandra_tests/validation/operations/select_test.py:1301]: Surprisingly, if I modify `test_local_indexed_values` to use value 65535, it errors out with:
```
E           AssertionError: Unexpected errors found:
E           node3: 1 errors
E           ERROR 2025-03-28 10:08:36,733 [shard 1:sl:d] storage_proxy - exception during mutation write to cbb7a005-8e1b-4b08-a010-e37337eb8453: std::runtime_error (Key size too large: 65537 > 65535)
```
As if we add some "metadata" (maybe type?) to the key. 
Not sure how to proceed, do we want to lower the acceptable limit by 2 then? Or more? I need to understand where this difference comes from first, I think, to make an informed decision.

Comment (ptrsmrn) [reply test/cqlpy/cassandra_tests/validation/operations/select_test.py:1301]: If I run this testcase on a pure `master` is errors out with:
```
    def testPKQueryWithValueOver64K(cql, test_keyspace):
        with create_table(cql, test_keyspace, "(a text, b text, PRIMARY KEY (a, b))") as table:
            TOO_BIG = 1024 * 65
            too_big = 'x'*TOO_BIG
>           assert_invalid_throw(cql, table, InvalidRequest,
                               "SELECT * FROM %s WHERE a = ?", too_big)

cassandra_tests/validation/operations/select_test.py:1303: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
cassandra_tests/porting.py:109: in assert_invalid_throw
    execute(cql, table, cmd, *args)
cassandra_tests/porting.py:93: in execute
    return cql.execute(prepared, args)
cassandra/cluster.py:2762: in cassandra.cluster.Session.execute
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   cassandra.cluster.NoHostAvailable: ('Unable to complete the operation against any hosts', {<Host: 127.2.40.147:9042 datacenter1>: <Error from server: code=0000 [Server error] message="Key size too large: 66562 > 65535">})

Comment (ptrsmrn) [reply test/cqlpy/cassandra_tests/validation/operations/select_test.py:1301]: > it tests LONG_TEXT_LENGTH = 8193 OVERSIZE_LENGTH = 66536
> 
> but not OVERSIZE_LENGTH-1

@nuivall note that 66536 is not 64KB=65536, IDK where this number 66536 comes from.

Comment (ptrsmrn) [reply test/cqlpy/cassandra_tests/validation/operations/select_test.py:1301]: I'm a bit lost: I'm not sure if this patch is needed, seems like we're already doing the key length checks (for both the INSERT and SELECT), it's just that we print a bit different error.

**Comment by mykaul** [comment]
@ptrsmrn - what's the next step here?

**Comment by ptrsmrn** [comment]
> @ptrsmrn - what's the next step here?

@mykaul it never got priority, it's an old issue nobody really cares about any longer.

**Comment by ptrsmrn** [comment]
> #### 🔴 CI State: FAILURE
> Mode	Stage	Status	Queue Time	Run Time	Node
> [Checkout](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/28714//pipeline-overview)	✅		2m	spider9.cloudius-systems.com
> [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/28714//pipeline-overview)	✅		0m	spider9.cloudius-systems.com
> [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/28714//pipeline-overview)	✅	5m	41m	spider17.cloudius-systems.com
> dev	[test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/28714//pipeline-overview)	✅		13m	spider17.cloudius-systems.com
> [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/28714//pipeline-overview)
> 🔹 test_insert_at_max_ck_plus_one_fails
> 🔹 test_insert_at_max_pk_plus_one_fails
> 🔹 test_select_at_max_pk_plus_one_fails
> 🔹 test/cluster/dtest/limits_test.py	❌	0m	46m	spider17.cloudius-systems.com
> #### Failed Tests (2/9719):
> * [test_max_cells.dev.13](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/28714/testReport/junit/cluster.dtest.limits_test/TestLimits/Unit_Tests_Custom___dev___test_max_cells_dev_13) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+test_max_cells.dev.13)
> * [test_max_cells.dev.71](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/28714/testReport/junit/cluster.dtest.limits_test/TestLimits/Unit_Tests_Custom___dev___test_max_cells_dev_71) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+test_max_cells.dev.71)
> 
> #### Build Details:
> * Total Queue: 5m
> * Total Run: 1h 47m
> 
> **Note:** To re-trigger CI for this PR, comment: `@scylladbbot trigger-ci`

by AI:
> Build succeeds. Now quick test run to confirm the relevant tests still pass:

so I guess it wasn't enough.

**Comment by ptrsmrn** [comment]
> > #### 🔴 CI State: FAILURE
> > Mode	Stage	Status	Queue Time	Run Time	Node
> > [Checkout](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/28714//pipeline-overview)	✅		2m	spider9.cloudius-systems.com
> > [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/28714//pipeline-overview)	✅		0m	spider9.cloudius-systems.com
> > [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/28714//pipeline-overview)	✅	5m	41m	spider17.cloudius-systems.com
> > dev	[test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/28714//pipeline-overview)	✅		13m	spider17.cloudius-systems.com
> > [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/28714//pipeline-overview)
> > 🔹 test_insert_at_max_ck_plus_one_fails
> > 🔹 test_insert_at_max_pk_plus_one_fails
> > 🔹 test_select_at_max_pk_plus_one_fails
> > 🔹 test/cluster/dtest/limits_test.py	❌	0m	46m	spider17.cloudius-systems.com
> > #### Failed Tests (2/9719):
> > 
> > * [test_max_cells.dev.13](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/28714/testReport/junit/cluster.dtest.limits_test/TestLimits/Unit_Tests_Custom___dev___test_max_cells_dev_13) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+test_max_cells.dev.13)
> > * [test_max_cells.dev.71](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/28714/testReport/junit/cluster.dtest.limits_test/TestLimits/Unit_Tests_Custom___dev___test_max_cells_dev_71) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+test_max_cells.dev.71)
> > 
> > #### Build Details:
> > 
> > * Total Queue: 5m
> > * Total Run: 1h 47m
> > 
> > **Note:** To re-trigger CI for this PR, comment: `@scylladbbot trigger-ci`
> 
> by AI:
> 
> > Build succeeds. Now quick test run to confirm the relevant tests still pass:
> 
> so I guess it wasn't enough.

Hmm but it passes for me locally... Let me restart the CI first, I'll continue investigating it locally it fails again, maybe my env is different. Otherwise I'll open a Jira.

**Comment by ptrsmrn** [comment]
@scylladbbot trigger-ci

**Comment by ptrsmrn** [comment]
Looking into CI failures...

**Comment by ptrsmrn** [comment]
> #### 🔴 CI State: FAILURE
> Mode	Stage	Status	Queue Time	Run Time	Node
> [Checkout](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/28726//pipeline-overview)	✅		2m	spider9.cloudius-systems.com
> [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/28726//pipeline-overview)	✅		0m	spider9.cloudius-systems.com
> [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/28726//pipeline-overview)	✅	1h 22m	43m	spider18.cloudius-systems.com
> dev	[test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/28726//pipeline-overview)	✅		13m	spider18.cloudius-systems.com
> [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/28726//pipeline-overview)
> 🔹 test_insert_at_max_ck_plus_one_fails
> 🔹 test_insert_at_max_pk_plus_one_fails
> 🔹 test_select_at_max_pk_plus_one_fails
> 🔹 test/cluster/dtest/limits_test.py	❌	0m	35m	spider18.cloudius-systems.com
> #### Failed Tests (3/9684):
> * [test_max_cells.dev.51](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/28726/testReport/junit/cluster.dtest.limits_test/TestLimits/Unit_Tests_Custom___dev___test_max_cells_dev_51) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+test_max_cells.dev.51)
> * [test_max_cells.dev.74](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/28726/testReport/junit/cluster.dtest.limits_test/TestLimits/Unit_Tests_Custom___dev___test_max_cells_dev_74) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+test_max_cells.dev.74)
> * [test_max_cells.dev.92](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/28726/testReport/junit/cluster.dtest.limits_test/TestLimits/Unit_Tests_Custom___dev___test_max_cells_dev_92) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+test_max_cells.dev.92)
> 
> #### Build Details:
> * Total Queue: 1h 22m
> * Total Run: 2h 58m
> 
> **Note:** To re-trigger CI for this PR, comment: `@scylladbbot trigger-ci`

Ah so it's a flaky test after all: https://scylladb.atlassian.net/browse/SCYLLADB-1645?search_id=8babbee1-72ce-41fe-ac9c-40f6ecb72e83&referrer=quick-find

**Comment by ptrsmrn** [comment]
@scylladbbot trigger-ci

File: keys/compound.hh [hunk 5]
```diff
@@ -103,7 +104,12 @@ public:
     static managed_bytes serialize_value(RangeOfSerializedComponents&& values) {
         auto size = serialized_size(values);
         if (size > std::numeric_limits<size_type>::max()) {
-            throw std::runtime_error(format("Key size too large: {:d} > {:d}", size, std::numeric_limits<size_type>::max()));
+            // Matches Cassandra's wording so CQL-level compatibility tests
+            // (and client-visible error messages) line up.
+            // Issues #10366 (SELECT) and #12247 (INSERT) both require a
+            // clean InvalidRequest here rather than a generic server error.
+            throw exceptions::invalid_request_exception(format("Key length of {:d} is longer than maximum of {:d}",
```
Comment (nyh) [inline keys/compound.hh]: I was sure serialize_value() was only used on the replica, but if it's used in the coordinator, everything becomes easier.

There's a bit of an ugliness here: keys/compound.hh is a low-level feature, it's not specific to CQL. So it's very strange this low-level code throws CQL-specific exceptions.

So I don't love this solution, but will not disqualify it either.

File: test/cqlpy/cassandra_tests/validation/operations/select_test.py (hunk 2 above)
Comment (nyh) [reply test/cqlpy/cassandra_tests/validation/operations/select_test.py:1309]: Yes, these tests are translated from Cassandra so we shouldn't modify except for some trivial things like slightly different error messages. 
If we decide this test actually enshrines a Cassandra **bug** and we don't want to enshrine, commenting out the test or skipping it (with a comment) is also possible. I think xfail is not the best option, because I don't think we'll ever want to be compatible with what looks like a Cassandra bug.

File: test/cqlpy/test_key_length.py [hunk 6]
```diff
@@ -87,6 +83,30 @@ def test_where_65k_ck(cql, table1):
...
+# Scylla because of the 2-byte framing overhead per component) must not
+# error out during SELECT; values at MAX-allowed+1 must raise a clean
+# InvalidRequest. The MAX-allowed case for the 65533/65535 discrepancy is
+# tracked by #16772 and remains xfail; here we just pin the failure side
+# of the boundary so a regression would be caught immediately.
+def test_select_at_max_pk_plus_one_fails(cql, table1):
+    stmt = cql.prepare(f'SELECT * FROM {table1} WHERE p = ?')
+    over = 'x' * (65535 + 1)
+    with pytest.raises(InvalidRequest, match='Key length'):
+        cql.execute(stmt, [over])
```
Comment (nyh) [inline test/cqlpy/test_key_length.py]: We already have exactly the same test test_insert_65k_pk for 65*1024 - and this test is for 64*1024. I think it's not really necessary to have both... But I guess it's not a disaster.

File: test/cqlpy/test_key_length.py [hunk 7]
```diff
@@ -137,8 +157,12 @@ def table2(cql, test_keyspace):
 # Cassandra also has a bug (see CASSANDRA-19270) in the compound pk case -
 # instead of the expected InvalidRequest error, it generates an internal
 # server error (i.e., NoHostAvailable) with the message "'H' format requires
-# 0 <= number <= 65535".
-@pytest.mark.xfail(reason="Issue #12247")
+# 0 <= number <= 65535". Scylla uses the same CQL wire protocol, so the
+# Python driver itself rejects the oversized component before it ever
+# reaches the server - we therefore cannot produce a clean InvalidRequest
```
Comment (nyh) [inline test/cqlpy/test_key_length.py]: I don't understand... The "H format" message came from Cassandra, not from Python... Or am I wrong?
If it's a Python problem, the whole test should be rethought - we don't want an "xfail" when it is a driver bug, not a Scylla bug. But is it a driver bug? I'm not convinced. Maybe you can show some evidence?

**Review by nyh — APPROVED** [review]
Looks mostly good, so I tentatively approve, but I did leave some comments/requests so please take a look.

File: keys/compound.hh (hunk 5 above)
Comment (ptrsmrn) [reply keys/compound.hh]: Alternative was to throw a domain-neutral exception (std::invalid_argument or a new keys::key_too_long) and translate it to invalid_request_exception at every CQL boundary that calls serialize_value, but that's many catch sites instead of just this one. Also, there's strong precedent in the codebase for other low-level layers (outside of cql3/) throwing exceptions::invalid_request_exception directly.

File: test/cqlpy/test_key_length.py (hunk 6 above)
Comment (ptrsmrn) [reply test/cqlpy/test_key_length.py]: They differ on purpose. test_insert_65k_pk uses 65*1024 = 66560, i.e. clearly over the limit, ~1KB above; it's a smoke test that "very large" is rejected. The new test_*_at_max_pk_plus_one_fails use exactly 65535+1 = 65536 to pin MAX-allowed + 1, the tight boundary. The pair was added in response to @nuivall's earlier review on this PR https://github.com/scylladb/scylladb/pull/23433#discussion_r2012863246, which asked for MAX allowed and MAX allowed + 1 cases. I left the existing 65*1024 tests in place because they reproduce the original bug reports (#10366, #12247) verbatim.

File: test/cqlpy/test_key_length.py (hunk 7 above)
Comment (ptrsmrn) [reply test/cqlpy/test_key_length.py]: My understanding is it's a driver issue: python driver computes the token-aware routing key client-side before sending the request. For a compound partition key it serializes each component with a 2-byte length prefix:
`cassandra/query.py:304`:
> yield struct.pack(">H%dsB" % l, l, p, 0)

When l > 65535, struct.pack raises struct.error("'H' format requires 0 <= number <= 65535"), which the driver wraps in a NoHostAvailable. Reproduced by running this very test with --runxfail against my fix:

 > cassandra.cluster.NoHostAvailable: ('Unable to complete the operation against any hosts',
>     {<Host: 127.15.3.1:9042 datacenter1>: error("'H' format requires 0 <= number <= 65535")})

Note the value in the dict is error(...) rather than <Error from server: ...> — i.e. it never reached Scylla. Server log confirms zero requests for the oversized key.

I've tightened the in-code comment to point directly at cassandra/query.py:304 (was vaguely referring to "CQL framing limit" before). The xfail stays — we'd need the driver to either lift this limit or surface it as InvalidRequest for the test to pass.

File: cql3/restrictions/statement_restrictions.cc (hunk 4 above)
Comment (ptrsmrn) [reply cql3/restrictions/statement_restrictions.cc]: Reworked.

File: test/cqlpy/cassandra_tests/validation/operations/select_test.py (hunk 3 above)
Comment (ptrsmrn) [reply test/cqlpy/cassandra_tests/validation/operations/select_test.py:1303]: I added the extra testcase requested in the 1st msg.

File: cql3/restrictions/statement_restrictions.cc (hunk 1 above)
Comment (ptrsmrn) [reply cql3/restrictions/statement_restrictions.cc]: No longer valid after rework, closing.

**Comment by ptrsmrn** [comment]
> @scylladbbot trigger-ci

I don't understand what happened with the CI, there was no report posted. But since I force pushed, I'm waiting for another CI run.

**Comment by nuivall** [comment]
there are some `@pytest.mark.xfail(reason="issue #12247")` left in test/cqlpy/cassandra_tests/validation/operations/insert_invalidate_sized_records_test.py

File: test/cqlpy/cassandra_tests/validation/operations/insert_invalidate_sized_records_test.py [hunk 8]
```diff
@@ -40,17 +42,17 @@ def testsingleValuePk(cql, test_keyspace):
...
             execute(cql, table, "INSERT INTO %s (a) VALUES (?)", b'')
 
-@pytest.mark.xfail(reason="issue #12247")
+@pytest.mark.xfail(reason="Driver rejects oversized compound-pk component (struct.pack '>H'); see CASSANDRA-19270, #12247")
 # Currently fails on Cassandra due to CASSANDRA-19270
 def testcompositeValuePk(cql, test_keyspace, cassandra_bug):
     with create_table(cql, test_keyspace, "(a blob, b blob, PRIMARY KEY ((a, b)))") as table:
         # sum of columns is too large
-        with pytest.raises(InvalidRequest, match=f'Key length of {len(MEDIUM_BLOB)*2} is longer than maximum of 65535'):
+        with pytest.raises(InvalidRequest, match=r'Key length of \d+ is longer than maximum of 65535'):
```
Comment (nuivall) [inline test/cqlpy/cassandra_tests/validation/operations/insert_invalidate_sized_records_test.py:50]: why change this if it xfails?

**Comment by ptrsmrn** [comment]
UPDATE:
- removed leftover xfail for #12247 in insert_invalidate_sized_records_test.py (per @nuivall)
- updated testcompositeValuePk xfail reason to driver-side rejection (CASSANDRA-19270) - otherwise it fails
- needed to relax exact-size regex to \d+ in that file
- testsingleValueIndex stays xfailed for #8627 (separate bug)

File: test/cqlpy/cassandra_tests/validation/operations/insert_invalidate_sized_records_test.py (hunk 8 above)
Comment (nuivall) [reply test/cqlpy/cassandra_tests/validation/operations/insert_invalidate_sized_records_test.py:50]: ok, maybe makes sense to align all tests

Comment (ptrsmrn) [reply test/cqlpy/cassandra_tests/validation/operations/insert_invalidate_sized_records_test.py:50]: Updating it now means future un-xfail won't be tripped up by a stale literal.

**Review by nuivall — APPROVED** [review]

**Comment by ptrsmrn** [comment]
@scylladbbot trigger-ci

**Comment by ptrsmrn** [comment]
Previously the CI failed on a known https://scylladb.atlassian.net/browse/SCYLLADB-1645 (the CI report was not posted under this PR).
